### PR TITLE
Make consensus layer compatible with GHC 8.10

### PR DESCRIPTION
**Pull request by mrBliss**
All changes are backwards-compatible with 8.6

* Replace `*` with `Type`
* Remove imports that GHC now considers redundant
* `fail` is no longer a method of `Monad`, but of `MonadFail`. See
  <https://gitlab.haskell.org/ghc/ghc/-/wikis/proposal/monad-fail#adapting-old-code>
* Remove record wildcard matches when no fields are
  used (`-Wunused-record-wildcards`)

### CODE REVIEW

**Comment by mrBliss** [comment]
I'm extracting this from #2539 because that PR currently has some test failures and I'd like to merge this ASAP to avoid merge conflicts.

File: ouroboros-consensus-cardano/tools/db-analyser/Analysis.hs [hunk 1]
```diff
@@ -215,7 +211,7 @@ processAllChainDB chainDB rr callback = do
       Origin -> return ()
       At tip -> do
         Right itr <- ChainDB.stream chainDB rr GetBlock
-          (StreamFromExclusive genesisPoint)
+          (StreamFromExclusive GenesisPoint)
```
Comment (nc6) [inline ouroboros-consensus-cardano/tools/db-analyser/Analysis.hs:214]: Is this change related?

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs [hunk 2]
```diff
@@ -13,6 +13,7 @@
 {-# LANGUAGE ScopedTypeVariables    #-}
 {-# LANGUAGE TypeApplications       #-}
 {-# LANGUAGE TypeFamilies           #-}
+{-# LANGUAGE UndecidableInstances   #-}
```
Comment (nc6) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs:16]: Hmm, where is this coming from? The migration notes for 8.10 say this might be needed where we have some injective type families, but I don't see that here...

**Review by nc6 — APPROVED** [review]
Looks good. Don't quite understand where the `UndecidableInstances` is coming from.

File: ouroboros-consensus-cardano/tools/db-analyser/Analysis.hs (hunk 1 above)
Comment (mrBliss) [reply ouroboros-consensus-cardano/tools/db-analyser/Analysis.hs:214]: `Ouroboros.Consensus.Block` re-exports most from `Ouroboros.Network.Block`, but not `genesisPoint`, only the `GenesisPoint` pattern synonym.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs (hunk 2 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs:16]: This is the error I get without it:
```
src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs:311:3: error:
    • Variable ‘c'’ occurs more often
        in the constraint ‘c'’ than in the instance head ‘c’
      (Use UndecidableInstances to permit this)
    • In the quantified constraint ‘c' => c’
      In the definition of data constructor ‘Weaken’
      In the data type declaration for ‘Ap’
    |
311 |   Weaken :: (c' => c) => Ap m l r b c -> Ap m l r b c'
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs:16]: It's kind of strange, because it's a GADT constructor, not an instance declaration (probably due to GHC's error reporting).

**Comment by mrBliss** [comment]
bors merge

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/SOP.hs [hunk 3]
```diff
@@ -128,7 +129,7 @@ lenses_NP = go sList
 
 -- | Conjure up an 'SListI' constraint from an 'NP'
 npToSListI :: NP a xs -> (SListI xs => r) -> r
-npToSListI = sListToSListI . npToSList
+npToSListI np = sListToSListI $ npToSList np
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Util/SOP.hs:132]: This change is due to impredicativity:
```
src/Ouroboros/Consensus/Util/SOP.hs:132:14: error:
    • Cannot instantiate unification variable ‘c0’
      with a type involving polytypes: (SListI xs => r0) -> r0
        GHC doesn't yet support impredicative polymorphism
    • In the first argument of ‘(.)’, namely ‘sListToSListI’
      In the expression: sListToSListI . npToSList
      In an equation for ‘npToSListI’:
          npToSListI
            = sListToSListI . npToSList
            where
                sListToSListI :: SList xs -> (SListI xs => r) -> r
                sListToSListI SNil k = k
                sListToSListI SCons k = k
                npToSList :: NP a xs -> SList xs
                npToSList Nil = SNil
                npToSList (_ :* xs) = sListToSListI (npToSList xs) SCons
    • Relevant bindings include
        npToSListI :: NP a xs -> (SListI xs => r) -> r
          (bound at src/Ouroboros/Consensus/Util/SOP.hs:132:1)
```

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/GcSchedule.hs [hunk 4]
```diff
@@ -8,7 +8,7 @@ module Test.Ouroboros.Storage.ChainDB.GcSchedule (tests, example) where
 import           Control.Monad (forM)
 import           Control.Tracer (nullTracer)
 import           Data.Fixed (div')
-import           Data.List
+import           Data.List (foldl', partition, sort)
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/GcSchedule.hs:11]:
```
test-storage/Test/Ouroboros/Storage/ChainDB/GcSchedule.hs:11:18: error: [-Wcompat-unqualified-imports, -Werror=compat-unqualified-imports]
    To ensure compatibility with future core libraries changes
    imports to Data.List should be
    either qualified or have an explicit import list.
   |
11 | import           Data.List
   |          
```

**Comment by mrBliss** [comment]
bors merge

**Comment by mrBliss** [comment]
bors merge

**Comment by mrBliss** [comment]
Forgot to mention: https://github.com/input-output-hk/ouroboros-network/pull/2540/commits/f07b7eef67ccd69171e1e88e927a4b154fab4ad6 will be needed to really make it compatible with 8.10, but I'd rather merge that after we have switched to 8.10.
